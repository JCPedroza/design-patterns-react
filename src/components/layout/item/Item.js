import { arrayOf, object, string } from 'prop-types'

const buildItemToShow = (item, keysToShow) => {
  const itemProxy = new Proxy(item, { ownKeys: () => keysToShow })
  return Object.entries(Object.assign({}, itemProxy))
}

const Item = ({ item, keysToShow }) => {
  const itemToShow = buildItemToShow(item, keysToShow)

  return (
    <>
      <ul>
        {itemToShow.map(([key, value]) => <li key={key}>{key}: {value}</li>)}
      </ul>
    </>
  )
}

Item.propTypes = {
  item: object.isRequired,
  keysToShow: arrayOf(string).isRequired
}

export default Item
