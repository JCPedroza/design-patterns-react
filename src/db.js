/**
 * Schema-less, no validation, no authentication, singleton mock database.
 */

/**
 * Data that the collection holds.
 */
const data = [
  { id: 14, title: 'One', author: 'Mario', year: 1926, time: 123 },
  { id: 38, title: 'Tuesday', author: 'Fernanda', year: 2013, time: 234 },
  { id: 23, title: 'Blue', author: 'Unknown', year: 1926, time: 428 },
  { id: 49, title: 'Untitled', author: 'Radiohead', year: 2003, time: 357 }
]

const isMatch = (target, filter) => {
  return Object.keys(filter).every(key => {
    if (typeof filter[key] === 'object') {
      return isMatch(target[key], filter[key])
    }

    return target[key] === filter[key]
  })
}

const getIndexOfId = (id) => {
  const parsedId = parseInt(id, 10)
  const index = data.findIndex(point => point.id === parsedId)
  if (parsedId === -1) throw new Error(`Cannot find id: ${id}`)
  return index
}

const injectData = (newData) =>
  data.splice(0, data.length, ...newData)

const find = (filter) =>
  filter ? data.filter(song => isMatch(song, filter)) : data

const getById = (id) =>
  data.find(song => song.id === id)

const create = (dataPoint) =>
  data[data.push(dataPoint) - 1]

const getByIdAndReplace = (id, replacement) =>
  data.splice(getIndexOfId(id), 1, replacement)[0]

const getByIdAndUpdate = (id, update) => {
  const index = getIndexOfId(id)
  data[index] = { ...data[index], ...update }
  return data[index]
}

const getByIdAndDelete = (id) =>
  data.splice(getIndexOfId(id), 1)[0]

const api = {
  injectData,
  find,
  getById,
  create,
  getByIdAndReplace,
  getByIdAndUpdate,
  getByIdAndDelete
}

export default api
