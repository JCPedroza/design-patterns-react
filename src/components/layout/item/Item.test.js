import { render } from '@testing-library/react'

import Item from './Item'

import db from '../../../db'

const itemData = { id: 1, title: 'One', author: 'Mario', year: 1926 }

jest.mock('../../../db', () => ({
  ...jest.requireActual('../../../db'),
  getById: jest.fn().mockReturnValue(itemData)
}))

const song = db.getById(1)
const keysToShow = ['title', 'year']

it('renders item', () => {
  const { container } = render(
    <Item item={song} keysToShow={keysToShow} />
  )

  expect(container).toMatchSnapshot()
})
