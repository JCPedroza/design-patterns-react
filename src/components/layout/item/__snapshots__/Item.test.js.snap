// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item component renders item data 1`] = `
<div>
  <ul>
    <li>
      title
      : 
      One
    </li>
    <li>
      year
      : 
      1926
    </li>
  </ul>
</div>
`;
